### DISCUSSION TITLE
Introducing Reaktoro-pse, a library for integrating Reaktoro into IDAES and WaterTAP equation-oriented modeling platform

### DISCUSSION THREAD

**Discussion by avdudchenko** [reaktoro/reaktoro#407] [Show and tell]
We wanted to share with you a recently developed [library for integrating Reaktoro as a Graybox model](https://github.com/watertap-org/reaktoro-pse) in [Pyomo](https://pyomo.readthedocs.io/en/stable/), for [IDAES]( https://idaes-pse.readthedocs.io/en/stable/)  compatible modeling libraries, with specific focus on  [WaterTAP](https://watertap.readthedocs.io/en/stable/).  

The objective of this Reaktoro-pse library is to provide simple method to use Reaktoro equilibrium solver and output properties in process models that are being solved using  [IPOPT]( https://cyipopt.readthedocs.io/en/stable/). This, for example, [enables the use of Reaktoro to estimate Osmotic pressure in Reverse osmosis model]( https://github.com/watertap-org/reaktoro-pse/blob/main/src/reaktoro_pse/tutorials/integration_with_ro.ipynb), [vapor pressures in distillation processes, and solid precipitation in a softening process]( https://github.com/watertap-org/reaktoro-pse/blob/main/src/reaktoro_pse/examples/thermal_precipitation.py). 
Of course, the application of Reaktoro-pse is only limited by the capabilities of Reaktoro itself, so in theory, any equilibrium calculation Reaktoro can perform is supported (kinetics are coming next), allowing for use in a variety of applications. 

Who is Reaktoro-pse for? 
(a)	Anyone trying to include Reaktoro calculation in a process-level model 
(b)	Anyone trying to include equations and constraints to configure Reaktoro inputs or use Reaktoro outputs that are not supported by Reaktoro natively (for example, performing Reaktoro calculations in systems with recycle loops)

We are happy to hear feedback from potential users, as well as requests for feature support. This is the first version of the API, which will see further improvements with active use. 

Finally, we want to thank @allanleal , for all the help, and rest of Reaktoro developers and contributors for creating Reaktoro, which we earnestly believe is a fantastic open-source tool that brings open-source chemistry modeling into 21st century.

**Comment by allanleal** [comment]
Hi @avdudchenko - fantastic news! Congratulations on reaching this milestone. I'm very happy to see your advanced use of Reaktoro, even relying on capabilities that are not yet documented. Well done!

I look forward to learning more about reaktoro-pse soon!

Thanks for sharing the good news.

**Comment by allanleal** [comment]
On a separate note, I would suggest that you choose the Pitzer activity model for the aqueous phase as default model, instead of ideal, given that the default thermodynamic database is `pitzer.dat`.

<img width="368" alt="image" src="https://github.com/user-attachments/assets/c7b578a2-c840-4f97-9e0e-33a27b662de7">

So, instead of `ActivityModelIdealAqueous`, you would use `ActivityModelPitzer`. But maybe you have good reasons for keeping the ideal as default.

**Comment by avdudchenko** [comment]
Thanks for taking a look. 

The reason for ideal activity models being default is really to force users to make an effort to learn reaktoro before using reaktoro-pse. 

Throughout all the examples, we show that the user should specify the Pitzer Activity model. However, maybe it makes more sense to pick more reasonable defaults than ideal activity models. This is something to think through as we refine the API and its structure.
